binary_search_tree/tree.py

```diff
@@ -14,35 +14,110 @@ class Tree:
+        return current_node 
+
     # Time Complexity: 
     # Space Complexity: 
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value) 
+        else:
+            self.add_helper(self.root, key, value)
 
     # Time Complexity: 
     # Space Complexity: 
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif  current.key >= key: 
+                current = current.left
+            else:
+                current = current.right
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # In-Order: Left, Current, Right
+    def inorder_helper(self, root, inorder_list):
+        if root:
+        # Traverse left
+            self.inorder_helper(root.left, inorder_list)
+        # Traverse root
+            # print(str(root.val) + "->", end='')
+            inorder_list.append({"key" : root.key, "value" : root.value})
+        # Traverse right
+            self.inorder_helper(root.right, inorder_list)
+        return inorder_list
+
+    # # Time Complexity: 
+    # # Space Complexity: 
     def inorder(self):
-        pass
+        inorder_list = []
+        if self.root == None:
+            return inorder_list
+        else:
+            return self.inorder_helper(self.root, inorder_list)
+
+    def preorder_helper(self, root, preorder_list):
+        if root:
+            preorder_list.append({"key" : root.key, "value" : root.value})
+            self.preorder_helper(root.left, preorder_list)
+            self.preorder_helper(root.right, preorder_list)
+        return preorder_list
 
     # Time Complexity: 
     # Space Complexity:     
     def preorder(self):
-        pass
+        preorder_list = []
+        if self.root == None:
+            return preorder_list
+        else:
+            return self.preorder_helper(self.root, preorder_list)
+
+    def postorder_helper(self, root, postorder_list):
+        if root:
+            self.postorder_helper(root.left, postorder_list)
+            self.postorder_helper(root.right, postorder_list)
+            postorder_list.append({"key" : root.key, "value" : root.value})
+        return postorder_list
 
     # Time Complexity: 
     # Space Complexity:     
     def postorder(self):
-        pass
+        postorder_list = []
+        if self.root == None:
+            return postorder_list
+        else:
+            return self.postorder_helper(self.root, postorder_list)
+
+
+    def height_helper(self, root):
+        if root == None:
+            return 0
+        lefth= self.height_helper(root.left)
+        righth = self.height_helper(root.right)
+
+        if lefth > righth:
+            return lefth + 1
+        else:
+            return righth + 1
 
     # Time Complexity: 
     # Space Complexity:     
```

Comment:
What do you think the space and time complexity of this is? Is the time complexity dependent on the size of the binary search tree? Is there any new variables being created that grow larger depending upon the input?